packages/react/slot/src/Slot.tsx

```diff
@@ -0,0 +1,100 @@
+import * as React from 'react';
+import { composeRefs, composeEventHandlers } from '@radix-ui/react-utils';
+
+/* -------------------------------------------------------------------------------------------------
+ * Slot
+ * -----------------------------------------------------------------------------------------------*/
+
+type SlotProps = { children: React.ReactNode };
+
+const Slot = React.forwardRef<never, SlotProps>((props, forwardedRef) => {
+  const { children, ...slotProps } = props;
+  const childLength = React.Children.count(children);
+
+  if (childLength === 1) {
+    return (
+      <SlotClone {...slotProps} ref={forwardedRef}>
+        {children}
+      </SlotClone>
+    );
+  }
+
+  return (
+    <>
+      {React.Children.map(children, (child) => {
+        if (React.isValidElement(child) && child.type === Slottable) {
+          return (
+            <SlotClone {...slotProps} ref={forwardedRef}>
+              {child.props.children}
+            </SlotClone>
+          );
+        }
+
+        return child;
+      })}
+    </>
+  );
+});
+
+Slot.displayName = 'Slot';
+
+/* -------------------------------------------------------------------------------------------------
+ * SlotClone
+ * -----------------------------------------------------------------------------------------------*/
+
+type SlotCloneProps = { children: React.ReactNode };
+
+const SlotClone = React.forwardRef<any, SlotCloneProps>((props, forwardedRef) => {
+  const { children, ...slotProps } = props;
+  const child = React.Children.only(children);
+
+  return React.isValidElement(child)
+    ? React.cloneElement(child, {
+        ...mergeProps(slotProps, child.props),
+        ref: composeRefs(forwardedRef, (child as any).ref),
+      })
+    : null;
+});
+
+SlotClone.displayName = 'SlotClone';
+
+/* -------------------------------------------------------------------------------------------------
+ * Slottable
+ * -----------------------------------------------------------------------------------------------*/
+
+const Slottable = ({ children }: { children: React.ReactNode }) => {
+  return children as React.ReactElement;
+};
+
+/* ---------------------------------------------------------------------------------------------- */
+
+function mergeProps<P extends { [key: string]: any }>(baseProps: P, otherProps: P) {
```

Comment:
I actually just thought, why do we need the generic? It made me think we're trying to infer something here but do we need inferrence? Would this suffice:

```suggestion
type AnyProps = Record<string, any>;
function mergeProps(baseProps: AnyProps, otherProps: AnyProps) {
```

Reply:
Yep, works, bonus we can remove the `as any`